### PR TITLE
Implement Unzippable

### PR DESCRIPTION
**Pull request by adamgfraser**
The dual of `Zippable`. Supports combining contravariant types in the same way as `Zippable` for covariant types. This allows us to implement compositional zips for `ZRef.Synchronized`.

### CODE REVIEW

**Review by jdegoes — DISMISSED** [review]

**Comment by jdegoes** [comment]
Clever idea. I like it!

File: core-tests/shared/src/test/scala/zio/ZRefSynchronizedSpec.scala [hunk 1]
```diff
@@ -490,7 +490,7 @@ object ZRefMSpec extends ZIOBaseSpec {
           tuple   <- composed.get
           (a, b)   = tuple
         } yield assert(a)(equalTo(6765)) && assert(b)(equalTo(10946))
-      },
+      } @@ nonFlaky,
```
Comment (jdegoes) [inline core-tests/shared/src/test/scala/zio/ZRefSynchronizedSpec.scala:493]: Is this strictly necessary?

**Review by jdegoes — APPROVED** [review]

File: core-tests/shared/src/test/scala/zio/ZRefSynchronizedSpec.scala (hunk 1 above)
Comment (adamgfraser) [reply core-tests/shared/src/test/scala/zio/ZRefSynchronizedSpec.scala:493]: Wanted to confirm that the updating of the two `Ref` values is performed atomically and we didn't just get "lucky" that two of the fibers updating the `Ref` didn't make conflicting updates, but happy to remove if you want.

**Comment by nartamonov** [comment]
> The dual of `Zippable`. Supports combining contravariant types in the same way as `Zippable` for covariant types. This allows us to implement compositional zips for `ZRef.Synchronized`.

Just discovered `Unzippable`, it's awesome! 👍

**Comment by adamgfraser** [comment]
🎉

**Pull request merged**